loader/loader_linux.c

```diff
@@ -0,0 +1,635 @@
+/*
+ *
+ * Copyright (c) 2021 The Khronos Group Inc.
+ * Copyright (c) 2021 Valve Corporation
+ * Copyright (c) 2021 LunarG, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Author: Mark Young <marky@lunarg.com>
+ *
+ */
+
+// Non-windows and non-apple only header file, guard it so that accidental
+// inclusion doesn't cause unknown header include errors
+#if defined(__linux__) || defined(__QNXNTO__) || defined(__FreeBSD__)
+
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "loader_linux.h"
+
+#include "allocation.h"
+#include "get_environment.h"
+#include "loader.h"
+#include "log.h"
+
+// Compare the two device types.
+// This behaves similar to a qsort compare.
+int32_t device_type_compare(VkPhysicalDeviceType a, VkPhysicalDeviceType b) {
```

Comment:
I feel like this would be simpler by using an integer and writing a three way comparison. Assign all the types a value (the same value for things which should compare the same). Then there is an absolute ordering for the preference.

This code to me looks very confusing, since theres a ton of if-else blocks and random 1's, 0's, and -1's.

Reply:
Ok, I'll clean it up